src/lib/util/vec.ml

```diff
@@ -28,144 +28,144 @@
+
+let[@inline] clear vec = vec.sz <- 0
+
+let[@inline] shrink vec i =
+  assert (i >= 0);
+  assert (i <= vec.sz);
+  for j = vec.sz - i to vec.sz - 1 do
+    Array.unsafe_set vec.data j vec.dummy
+  done;
+  vec.sz <- vec.sz - i
+
+let[@inline] pop vec =
+  assert (vec.sz > 0);
+  let x = Array.unsafe_get vec.data (vec.sz - 1) in
+  Array.unsafe_set vec.data (vec.sz - 1) vec.dummy;
+  vec.sz <- vec.sz - 1;
+  x
+
+let[@inline] last vec =
+  assert (vec.sz > 0);
+  Array.unsafe_get vec.data (vec.sz - 1)
+
+let[@inline] size vec = vec.sz
+
+let[@inline] is_empty vec = vec.sz = 0
+
+let[@inline] is_full vec = Array.length vec.data = vec.sz
+
+let[@inline] copy vec : _ t =
+  let data = Array.copy vec.data in
+  {vec with data}
+
+(* grow the array *)
+
+let[@inline never] grow_to vec cap : unit =
+  assert (Array.length vec.data < Sys.max_array_length);
+  let cap =
+    min Sys.max_array_length (max 4 cap)
+  in
+  let arr' = Array.make cap vec.dummy in
+  Array.blit vec.data 0 arr' 0 (Array.length vec.data);
+  vec.data <- arr';
+  assert (Array.length vec.data > vec.sz);
+  ()
+
+let[@inline never] grow_to_double_size vec : unit =
+  grow_to vec (2 * Array.length vec.data)
+
+let[@inline never] grow_to_by_double vec cap =
+  let cap = max 1 cap in
+  let c = ref (max 1 (Array.length vec.data)) in
+  while !c < cap do c := 2 * !c done;
+  grow_to vec !c
+
+let[@inline] push vec x : unit =
+  if is_full vec then grow_to_double_size vec;
+  Array.unsafe_set vec.data vec.sz x;
+  vec.sz <- vec.sz + 1
+
+let[@inline] get vec i =
+  assert (0 <= i && i < vec.sz);
+  let res = Array.unsafe_get vec.data i  in
+  if res == vec.dummy then raise Not_found;
+  res
+
+let[@inline] set vec i elt =
+  assert (not (elt == vec.dummy));
+  vec.data.(i) <- elt;
+  vec.sz <- max vec.sz (i+1)
+
+let[@inline] fast_remove vec i =
+  assert (i>= 0 && i < vec.sz);
+  Array.unsafe_set vec.data i @@ Array.unsafe_get vec.data (vec.sz - 1);
+  vec.sz <- vec.sz - 1
+
+let filter_in_place f vec =
+  let i = ref 0 in
+  while !i < size vec do
+    if f (Array.unsafe_get vec.data !i) then incr i else fast_remove vec !i
+  done
 
-let from_array data sz d = {data = data; sz = sz; dummy = d}
+let sort vec f : unit =
+  let sub_arr =
+    if is_full vec then vec.data else Array.sub vec.data 0 vec.sz
+  in
+  Array.fast_sort f sub_arr;
+  vec.data <- sub_arr
 
-let from_list l sz d =
-  let l = ref l in
-  let f_init _ = match !l with [] -> assert false | e::r -> l := r; e in
-  {data = Array.init sz f_init; sz = sz; dummy = d}
+let[@inline] iteri f vec =
+  for i = 0 to size vec - 1 do
+    let elt = Array.unsafe_get vec.data i in
+    if not (elt == vec.dummy) then
+      f i elt
+  done
 
-let clear s = s.sz <- 0
+let[@inline] iter f = iteri (fun _ elt -> f elt)
 
-let shrink t i fill_with_dummy =
-  assert (i >= 0 && i<=t.sz);
-  if fill_with_dummy then
-    for i = t.sz - i to t.sz - 1 do
-      t.data.(i) <- t.dummy
+let exists p vec =
+  let exception Terminate in
```

Comment:
Is there a reason to use a local exception here? If not it is slightly more efficient to declare the exception outside of `exists` (I know I am guilty of doing this as well, but `let exception` registers a new exception each time `exists` is called I believe)